File: Lab6/C++/Problem.cpp
#include "Problem.h"
#include <math.h>
#include <iostream>
#include <stdexcept>
using namespace std;
Problem::Problem(double a,double b,double c ) {
	this->a = a;
	this->b = b;
	this->c = c;
}
Problem::Problem() {

}
void Problem::setA(double a) {
	this->a = a;
}
void Problem::setB(double b) {
	this->b = b;
}
void Problem::setC(double c) {
	this->c = c;
}
void Problem::setD(double d) {
	this->d = d;
}
double Problem::getA() {
	return this->a;
}
double Problem::getB() {
	return this->b;
}
double Problem::getC() {
	return this->c;
}
double Problem::getD() {
	return this->d;
}
double operator/(Problem& p, double d) {
	double res = 0;
	try {
		res = p.Division(p.c,d);
	}
	catch (runtime_error& e) {
		cout << "Exception occurred" << endl
			<< e.what();
	}
	return res;
}
double Problem::ReturnLog10() {
	double res = 0;
	try {
		res = CheckArgument(4 * b - c);
	}
	catch (runtime_error & e) {
		cout<< "Exception occurred" << endl
			<< e.what();
	}
	return log10(4 * b - c);
}
double Problem::Multiply() {
	double res= ReturnLog10() * a;
	return res;
}
double operator*(Problem& p,double d) {
	double res = 0;
	try {
		res = p.Multiply() * p.Division(1 ,( (p.b + p.Division(p.c,d) - 1)));
	}
	catch (runtime_error& e) {
		cout << "Exception occurred" << endl
			<< e.what();
	}
	return res;
}
double Problem::Division(double c,double d) {
	//using stdexcept
	if(d==0)throw runtime_error("Math error: Attempted to divide by Zero\n");
	return c / d;
}
class Exception :public runtime_error {
public:
	Exception() :runtime_error("Math error: Log10 must be positive\n") {}
};

bool Problem::CheckArgument(double arg) {
	//using class exception
	if (arg <= 0) throw Exception();
	return true;
	
}

File: Lab6/C#/Problem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OOP_LAB6
{
    class Problem
    {
        private double a;
        private double b;
        private double c;
        private double d;
        public double A { get => a; set => a = value; }
        public double B { get => b; set => b = value; }
        public double C { get => c; set => c = value; }
        public double D { get => d; set => d = value; }
        public Problem(double a,double b,double c)
        {
            this.a = a;
            this.b = b;
            this.c = c;
            
        }
        public Problem()
        {

        }
        private double Lg()
        {
            double p = 0d;     
            try
            {
                p = 4 * b - c;
                if (p <=0)
                    throw new ArgumentException();
            }
            catch(Exception exp)
            {
                Console.WriteLine($"Argument of log10  must be positive.Error: {exp.Message}");
            }
            return Math.Log10(p);
        }
        
       private double Multiply()
        {
            return Lg() * a;
        }
        public static double operator /(Problem val1,double val2)
        {
            var t = 0d;
            try
            {
                
                if (val2 == 0 )
                    throw new DivideByZeroException();
                t = val1.c / val2;
            }
            catch (DivideByZeroException exp)
            {
                Console.WriteLine($"Error occured: {exp.Message}");
                
            }

            return t;
        }
        public static double operator *(Problem val1,double val2)
        {
            var t = 0d;
            if (val1 == null) val1 = new Problem();
            try
            {
                
               if ((val1.b + val1 / val2 - 1) == 0)
                    throw new DivideByZeroException();
                t = val1.Multiply() * (1 / (val1.b + val1 / val2 - 1));
            }
            catch (DivideByZeroException exp)
            {
                Console.WriteLine($"Error occured: {exp.Message}");
            }
            catch(NullReferenceException exp)
            {
                Console.WriteLine($"Error occured: {exp.Message}");
            }
            return t;
        }
       
       
    }
}
